File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/JsonConverter.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresReader;
import org.revenj.serialization.Serialization;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class JsonConverter {

	private final Serialization<String> serialization;

	public JsonConverter(Serialization<String> serialization) {
		this.serialization = serialization;
	}

	private String toJson(Map<String, Object> value) {
		if (value.isEmpty()) return "{}";
		try {
			return serialization.serialize(Map.class, value);
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		}
	}

	public void serializeURI(PostgresBuffer sw, Map<String, Object> value) {
		if (value == null) return;
		StringConverter.serializeURI(sw, toJson(value));
	}

	public void serializeCompositeURI(PostgresBuffer sw, Map<String, Object> value) {
		if (value == null) return;
		String str = toJson(value);
		StringConverter.serializeCompositeURI(sw, str);
	}

	public Map<String, Object> parse(PostgresReader reader, int context, boolean allowNulls) throws IOException {
		String input = StringConverter.parse(reader, context, true);
		if (input == null) {
			return allowNulls ? null : new HashMap<>(0);
		}
		return (Map<String, Object>) serialization.deserialize(Map.class, input);
	}

	public List<Map<String, Object>> parseCollection(
			PostgresReader reader,
			int context,
			boolean allowNulls) throws IOException {
		List<String> parsed = StringConverter.parseCollection(reader, context, true);
		if (parsed == null) return null;
		List<Map<String, Object>> list = new ArrayList<>(parsed.size());
		for (String s : parsed) {
			if (s == null) {
				if (allowNulls) list.add(null);
				else list.add(new HashMap<>(0));
			} else {
				list.add((Map<String, Object>) serialization.deserialize(Map.class, s));
			}
		}
		return list;
	}

	private static final PostgresTuple EMPTY = ValueTuple.from("{}");

	public PostgresTuple toTuple(Map<String, Object> value) {
		if (value == null) return null;
		else if (value.isEmpty()) return EMPTY;
		try {
			return ValueTuple.from(serialization.serialize(value));
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		}
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/JsonConverter.scala
package net.revenj.database.postgres.converters

import java.io.IOException

import net.revenj.database.postgres.{PostgresBuffer, PostgresReader}
import net.revenj.serialization.Serialization

class JsonConverter(serialization: Serialization[String]) extends Converter[Map[String, Any]] {
  override def serializeURI(sw: PostgresBuffer, value: Map[String, Any]): Unit = {
    if (value.isEmpty) {
      sw.addToBuffer("{}")
    } else {
      StringConverter.serializeURI(sw, serialization.serialize(value).getOrElse(throw new RuntimeException(s"Unable to serialize provided map: $value")))
    }
  }

  val dbName = "jsonb"

  def default() = Map.empty

  override def parseRaw(reader: PostgresReader, start: Int, context: Int): Map[String, Any] = {
    toMap(StringConverter.parseRaw(reader, start, context))
  }

  private def toMap(value: String) = {
    serialization.deserialize[Map[String, Any]](value).getOrElse(throw new RuntimeException("Unable to deserialize JsonB into map"))
  }

  override def parseCollectionItem(reader: PostgresReader, context: Int): Map[String, Any] = {
    val value = StringConverter.parseCollectionItem(reader, context)
    if (value.isEmpty) Map.empty
    else toMap(value)
  }

  override def parseNullableCollectionItem(reader: PostgresReader, context: Int): Option[Map[String, Any]] = {
    StringConverter.parseNullableCollectionItem(reader, context) match {
      case Some(value) =>
        if (value.isEmpty) Some(Map.empty)
        else Some(toMap(value))
      case _ => None
    }
  }

  private val EmptyTuple: PostgresTuple = ValueTuple.from("{}")

  override def toTuple(value: Map[String, Any]): PostgresTuple = {
    if (value.isEmpty) {
      EmptyTuple
    } else {
      ValueTuple.from(serialization.serialize(value).getOrElse(throw new IOException(s"Unable to serialize provided map: $value")))
    }
  }
}
